presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/WindowFilterPushDown.java

```diff
@@ -80,12 +80,18 @@ public PlanNode visitWindow(WindowNode node, RewriteContext<Constraint> context)
                 PlanNode rewrittenSource = context.rewrite(node.getSource(), null);
                 Optional<Integer> limit = getLimit(node, context.get());
                 if (node.getOrderBy().isEmpty()) {
-                    return new RowNumberNode(idAllocator.getNextId(),
+                    PlanNode rowNumberNode =  new RowNumberNode(idAllocator.getNextId(),
                             rewrittenSource,
                             node.getPartitionBy(),
                             getOnlyElement(node.getWindowFunctions().keySet()),
                             limit,
                             Optional.empty());
+                    if (limit.isPresent()) {
```

Comment:
This limit is not only generated by limit clause, but also the filter expression like `where rn <= 5`
And the ```LimitNode``` should only be added when limit clause is present.